src/execution/mod.rs

```diff
@@ -278,29 +338,53 @@ where
                             eof = true;
                         }
                         Poll::Ready(Some(item)) => {
-                            output.session(&cap).give(item);
+                            output_wrapper.activate().session(&output_cap).give(item);
                         }
                     }
                 }
             }
 
             if eof {
-                cap = None;
+                output_cap = None;
+                state_update_cap = None;
             } else {
                 activator.activate();
             }
         }
-    })
+    });
+
+    (output_stream, state_update_stream)
+}
+
+///
+fn resume_input_state(
+    worker_index: usize,
+    _worker_count: usize,
+    mut step_to_resume_state_bytes: HashMap<StateKey, StateBytes>,
+) -> (Option<StateBytes>, StateKey) {
+    let key = StateKey::Worker(worker_index);
+
+    let resume_state_bytes = step_to_resume_state_bytes.remove(&key);
+
+    (resume_state_bytes, key)
 }
 
 fn build_input_source<S>(
     py: Python,
     epoch_config: Py<EpochConfig>,
     scope: &S,
+    step_id: StepId,
+    key: StateKey,
```

Comment:
I'm interested in calling this `state_key`

Reply:
I'd like to refactor this later to more of the builder pattern so each source is in control of the key fully round-tripped through the recovery store and we don't actually have to pass it in here.